ASAN_SEGV | WebCore::RenderFragmentedFlow::objectShouldFragmentInFlowFragment.

lericaa opened the pull request:
#### 8251b5c26b6d9ddd3ac1268c2d8c7940094a49af
<pre>
ASAN_SEGV | WebCore::RenderFragmentedFlow::objectShouldFragmentInFlowFragment.
<a href="https://rdar.apple.com/125183911">rdar://125183911</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=272301">https://bugs.webkit.org/show_bug.cgi?id=272301</a>

Reviewed by Alan Baradlay.

This is another case where we can not dectect containing block change
in styledid/willchange like what was fixed in 266309@main,
moved markRendererDirtyAfterTopLayerChange to Element as common API
which can be used by FullscreenManager and HTMLDialogElement.

* LayoutTests/fast/css-grid-layout/out-of-flow-positioned-dialog-showModal-crash-expected.txt: Added.
* LayoutTests/fast/css-grid-layout/out-of-flow-positioned-dialog-showModal-crash.html: Added.
* Source/WebCore/dom/FullscreenManager.cpp:
(WebCore::FullscreenManager::willEnterFullscreen):
(WebCore::markRendererDirtyAfterTopLayerChange): Deleted.
* Source/WebCore/html/HTMLDialogElement.cpp:
(WebCore::HTMLDialogElement::showModal):
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::gridStyleHasNotChanged): mini refactoring to make criteria looks more clear.
(WebCore::RenderBox::updateGridPositionAfterStyleChange):
* Source/WebCore/rendering/RenderElement.cpp:
(WebCore::RenderElement::markRendererDirtyAfterTopLayerChange):
* Source/WebCore/rendering/RenderElement.h:

Canonical link: <a href="https://commits.webkit.org/278372@main">https://commits.webkit.org/278372@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1769d45dd7ebc235fc3ee1345711dc49a33ade95

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/50178 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/29469 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/2471 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/53436 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/868 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/52481 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/35695 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/594 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/40946 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/52277 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/27145 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/43184 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/22044 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/24568 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/435 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/8556 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/46555 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/496 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/55019 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/25274 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/516 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/48345 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/26535 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/43369 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/47359 "Passed tests") | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/11045 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/27398 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/26267 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

alanbaradlay commented on LayoutTests/fast/css-grid-layout/grid-crash-out-of-flow-positioned-dialog.html:
```diff
@@ -0,0 +1,23 @@
+<body>
+    <dialog id="dialog">
+    </dialog>
+</body>
```
```
<style>
html {
  position: absolute;
}
dialog {
  position: absolute;
  display: block;
}
</style><dialog id=dialog></dialog><script>
document.body.offsetHeight;
dialog.showModal();
</script>

```

test reduction.

alanbaradlay commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -402,18 +412,13 @@ void RenderBox::updateGridPositionAfterStyleChange(const RenderStyle& style, con
-    if (oldStyle->hasOutOfFlowPosition() && style.hasOutOfFlowPosition())
+    if ((oldStyle->hasOutOfFlowPosition() && style.hasOutOfFlowPosition()) || gridStyleHasNotChanged(style, oldStyle)) {
+        if (needsLayout() && isRenderGrid())
+            clearNeedsLayout();
```
1. needs layout bit should only be cleared when we actually finished running layout on the renderer
2. while out-of-flow boxes do not participate in grid layout, they do participate in an ancestor _block_ layout . Only when we run _that_ layout, should this bit be cleared.
3. this bug is not specific to grid (see test case above)

alanbaradlay replied:
1. when the _dialog_ becomes modal, it gets "anchored" to the iCB -i.e. the containing block of such box is always the RenderView.
2. Containing block type of changes are normally driven by style mutations (e.g. `position: static` -> `position: fixed`)
3. Such changes are handled in styleWillChange/didChange where we mark all affected ancestors dirty

Now `dialog.showModal()` normally goes through the same codepath triggered by this selector in the UA stylesheet
```
dialog:modal {
    position: fixed;

```
unless there's a user set value on position (note the missing `!important`) -which is what we have here.
In such cases we don't see any position type of style change and we fail to mark the ancestors dirty (which means that the subsequent layout does not reach the dialog box -> stays dirty)

There are 2 options I can think of
- force `position: fixed` on elements in top layer (should check the spec if it's doable at all)
- mark renderers dirty where we can actually detect this change e.g.

```
@ -4014,6 +4014,7 @@ void RenderLayer::establishesTopLayerWillChange()
 
     if (auto* parentLayer = parent())
         parentLayer->removeChild(*this);
+    renderer().setNeedsLayout();
 }
```

lericaa replied:
Thanks, for the detailed explanation and solutions. I will look into this issue again with all these info you provided.

lericaa replied:
The reduced test case helps me focus on real issue. Correct me if i am wrong, I think there is something else from the back trace i observed so far: 

At the beginning i was so confused why would #2 option work, as there will be a styleWillChange/didChange on `dialog.showModal()` anyway (it produces a `StyleDifference::Layout` diff which will  `setNeedsLayoutAndPrefWidthsRecalc();`). 

After compared the back trace between before and after applied #2, i found that the difference is that, this `setNeedsLayoutAndPrefWidthsRecalc();` is happened within `RenderElement::styleDidChange` which is happened after `styleWillChange` (where the original style bits will be cleared) and before `updateFromStyle();` (where style bits restored). this will cause an issue within `RenderObject::markContainingBlocksForLayout`, that `hasOutOfFlowPosition` check is always false. So `ancestor->setPosChildNeedsLayoutBit(true);` did not happen during. This is also the reason why #2 works and eventually the render layout got handled. 

For this specific case, now I feel this is more like a gap for OOF position when we set layout dirty bit during styleDidChange, and we expecting this render element will be handled during layoutBlock. I am wondering should we take its old style into consideration during `setNeedsLayoutAndPrefWidthsRecalc()->markContainingBlocksForLayout();` if it is done in `styleDidChange` to see this renderElement was `wasOutOfFlowPositioned`? Or am I wrong that is desired behaviour?

```
if (diff == StyleDifference::Layout || diff == StyleDifference::SimplifiedLayout) {
        RenderCounter::rendererStyleChanged(*this, oldStyle, m_style);

        // If the object already needs layout, then setNeedsLayout won't do
        // any work. But if the containing block has changed, then we may need
        // to mark the new containing blocks for layout. The change that can
        // directly affect the containing block of this object is a change to
        // the position style.
        if (needsLayout() && oldStyle && oldStyle->position() != m_style.position())
            markContainingBlocksForLayout();

        if (diff == StyleDifference::Layout)
->          setNeedsLayoutAndPrefWidthsRecalc();
        else
            setNeedsSimplifiedNormalFlowLayout();
```

Besides above, #2 has below code path that current code doesnt: 
```
void RenderObject::markContainingBlocksForLayout(ScheduleRelayout scheduleRelayout, RenderElement* newRoot)
{
...
if (hasOutOfFlowPosition) {
            bool willSkipRelativelyPositionedInlines = !ancestor->isRenderBlock() || ancestor->isAnonymousBlock();
            // Skip relatively positioned inlines and anonymous blocks to get to the enclosing RenderBlock.
            while (ancestor && (!ancestor->isRenderBlock() || ancestor->isAnonymousBlock()))
                ancestor = ancestor->container();
            if (!ancestor || ancestor->posChildNeedsLayout())
                return;
            if (willSkipRelativelyPositionedInlines)
                container = ancestor->container();
->          ancestor->setPosChildNeedsLayoutBit(true);
            simplifiedNormalFlowLayout = true;
```

alanbaradlay replied:
#2 works because we mark the containing block chain dirty _before_ the element gets added to the top layer. Soon after it is part of the top layer, any function that tries to climb the ancestor chain is going to get the "anchored to the ICB" answer because both `containingBlock()` and `container()` functions consult that bit ignoring the position property or any other bit on the renderer.

lericaa replied:
Got it, so this is expected. you also mentioned below which I did not quite understand previously. 
"unless there's a user set value on position (note the missing !important) -which is what we have here.
In such cases we don't see any position type of style change and we fail to mark the ancestors dirty"

Btw if we are going to apply #2 should we limited to isModal as popover will also have this code path if I remember correctly.

alanbaradlay commented on Source/WebCore/dom/Element.h:
```diff
@@ -617,6 +618,8 @@ class Element : public ContainerNode {
     void addToTopLayer();
     void removeFromTopLayer();
 
+    void markRendererDirtyAfterTopLayerChange(RenderElement* renderer, RenderBlock* containingBlockBeforeStyleResolution);
```
I think this should just be a helper function somewhere (and not an Element API)

lericaa replied:
Agreed that this is a helper function, just feel a bit odd to place it in either `HTMLDialogElement` or `FullscreenManager`. i updated patch to move it to RenderElement instead, let me know if it is appropriate.

alanbaradlay commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -391,6 +391,16 @@ void RenderBox::styleDidChange(StyleDifference diff, const RenderStyle* oldStyle
+        && oldStyle->order() == style.order()
+        && oldStyle->hasOutOfFlowPosition() == style.hasOutOfFlowPosition());
+}
+
```
any particular reason for making it a member function on RenderBox?

lericaa replied:
yeah, it should be a private helper function.

lericaa committed:
ASAN_SEGV | WebCore::RenderFragmentedFlow::objectShouldFragmentInFlowFragment.
rdar://125183911
https://bugs.webkit.org/show_bug.cgi?id=272301

Reviewed by Alan Baradlay.

This is another case where we can not dectect containing block change
in styledid/willchange like what was fixed in 266309@main,
moved markRendererDirtyAfterTopLayerChange to Element as common API
which can be used by FullscreenManager and HTMLDialogElement.

* LayoutTests/fast/css-grid-layout/out-of-flow-positioned-dialog-showModal-crash-expected.txt: Added.
* LayoutTests/fast/css-grid-layout/out-of-flow-positioned-dialog-showModal-crash.html: Added.
* Source/WebCore/dom/FullscreenManager.cpp:
(WebCore::FullscreenManager::willEnterFullscreen):
(WebCore::markRendererDirtyAfterTopLayerChange): Deleted.
* Source/WebCore/html/HTMLDialogElement.cpp:
(WebCore::HTMLDialogElement::showModal):
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::gridStyleHasNotChanged): mini refactoring to make criteria looks more clear.
(WebCore::RenderBox::updateGridPositionAfterStyleChange):
* Source/WebCore/rendering/RenderElement.cpp:
(WebCore::RenderElement::markRendererDirtyAfterTopLayerChange):
* Source/WebCore/rendering/RenderElement.h:

Canonical link: https://commits.webkit.org/278372@main
